lib/core/src/persist/sync.rs

```diff
@@ -0,0 +1,239 @@
+use anyhow::Result;
+use rusqlite::{named_params, Row};
+
+use super::Persister;
+use crate::sync::model::{sync::Record, SyncOutgoingDetails, SyncSettings, SyncState};
+
+impl Persister {
+    fn select_sync_state_query(where_clauses: Vec<String>) -> String {
+        let mut where_clause_str = String::new();
+        if !where_clauses.is_empty() {
+            where_clause_str = String::from("WHERE ");
+            where_clause_str.push_str(where_clauses.join(" AND ").as_str());
+        }
+
+        format!(
+            "
+            SELECT 
+                data_id,
+                record_id,
+                record_revision
+            FROM sync_state
+            {where_clause_str}
+        "
+        )
+    }
+
+    fn sql_row_to_sync_state(row: &Row) -> rusqlite::Result<Option<SyncState>> {
+        Ok(match row.get(0)? {
+            Some(data_id) => Some(SyncState {
+                data_id,
+                record_id: row.get(1)?,
+                record_revision: row.get(2)?,
+            }),
+            None => None,
+        })
+    }
+
+    pub(crate) fn get_sync_state_by_record_id(&self, record_id: &str) -> Result<Option<SyncState>> {
+        let con = self.get_connection()?;
+        let query = Self::select_sync_state_query(vec!["record_id = ?1".to_string()]);
+        let sync_state = con.query_row(&query, [record_id], Self::sql_row_to_sync_state)?;
+        Ok(sync_state)
+    }
+
+    pub(crate) fn get_sync_state_by_data_id(&self, data_id: &str) -> Result<Option<SyncState>> {
+        let con = self.get_connection()?;
+        let query = Self::select_sync_state_query(vec!["data_id = ?1".to_string()]);
+        let sync_state = con.query_row(&query, [data_id], Self::sql_row_to_sync_state)?;
+        Ok(sync_state)
+    }
+
+    pub(crate) fn set_sync_state(&self, sync_state: SyncState) -> Result<()> {
+        let con = self.get_connection()?;
+
+        con.execute(
+            "
+            INSERT OR REPLACE INTO sync_state(data_id, record_id, record_revision)
+            VALUES (:data_id, :record_id, :record_revision)
+        ",
+            named_params! {
+                ":data_id": &sync_state.data_id,
+                ":record_id": &sync_state.record_id,
+                ":record_revision": &sync_state.record_revision,
+            },
+        )?;
+
+        Ok(())
+    }
+
+    pub(crate) fn get_sync_settings(&self) -> Result<Option<SyncSettings>> {
+        let con = self.get_connection()?;
+
+        let sync_settings = con.query_row(
+            "
+            SELECT 
+                remote_url,
+                latest_revision
+            FROM sync_settings
+            WHERE id = 1
```

Comment:
If we always use 1 here then it seems we intend to keep only one record in this table.
Perhaps key/value columns better fit here?